pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+	}
+
+	p.requestID++
+
+	headers := http.Header{}
+	headers.Set(corev1.PortHeader, strconv.Itoa(port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(p.requestID))
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapConn(network, address, dataStream, errorStream), nil
+}
+
+// DialContext is a simple wrapper around Dial.
+func (p *PodDialer) DialContext(_ context.Context, network string, address string) (net.Conn, error) {
+	return p.Dial(network, address)
+}
+
+// parseDNS attempts to determine the intended pod to target, currently the
+// following formats are supported:
+//  1. [pod-name].[service-name].[namespace-name].svc.[cluster-domain]
+//  2. [pod-name].[service-name].[namespace-name].svc
+//  3. [pod-name].[service-name].[namespace-name]
+//  4. [pod-name].[service-name]
+//
+// If no cluster-domain is supplied, the dialer's configured domain is assumed.
+// If no namespace-name is supplied, the default namespace is assumed.
+// No validation is done to ensure that the pod is actually referenced by the
+// given service or that the DNS record exists in Kubernetes, instead this assumes
+// that things are set up properly in Kubernetes such that the FQDN passed here
+// matches Kubernetes' own DNS records and a pod within the cluster would be able
+// to resolve the same FQDN to the pod that we do.
+//
+// These assumptions allow us to use this dialer to issues requests *as if* we are
+// in the Kubernetes network even from outside of it (i.e. in tests that attempt to
+// connect to a pod at a given DNS address).
+func (p *PodDialer) parseDNS(fqdn string) (pod types.NamespacedName, port int, err error) {
+	addressPort := strings.Split(fqdn, ":")
+	if len(addressPort) != 2 {
+		err = ErrNoPort
+		return
+	}
+
+	port, err = strconv.Atoi(addressPort[1])
+	if err != nil {
+		return
+	}
+
+	fqdn = addressPort[0]
+	fqdn = strings.TrimSuffix(fqdn, "."+p.clusterDomain)
+	fqdn = strings.TrimSuffix(fqdn, "."+svcLabel)
+	labels := strings.Split(fqdn, ".")
+
+	// since we only dial pods we require 2 labels
+	// (assuming we're trying to dial a pod in the
+	// default namespace) or 3 (for a pod outside
+	// of the default namespace)
+	switch len(labels) {
+	case 2:
+		pod.Namespace = defaultNamespace
+	case 3:
+		pod.Namespace = labels[2]
+	default:
+		err = ErrInvalidPodFQDN
+		return
+	}
+
+	// service - labels[1]
+	pod.Name = labels[0]
+	return
+}
+
+// mutex must be held while calling this
+func (p *PodDialer) connectionForPod(pod types.NamespacedName) (httpstream.Connection, error) {
+	if conn, ok := p.connections[pod]; ok {
+		return conn, nil
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(p.config)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := p.config
```

Comment:
nit: It's not safe (I think) to copy rest configs in this manor. Prefer to use `DeepCopy` (or educate me if I'm mistaken I just checked if the method existed)

Reply:
I believe `*rest.Config` is fine to pass around, the `DeepCopy` that it has it actually inherits from embedding `*rest.TLSClientConfig`, so calling `DeepCopy` will return that rather than a `*rest.Config` - I believe you'd only have a problem if someone is modifying the config after you're passing it around due to synchronization issues.